### ISSUE TITLE
Member expiration calculation incorrect sometimes

### ISSUE THREAD

**Issue by carter2422** [restrictcontentpro/restrict-content-pro#299]
I had a user reactivate an expired account, yet their expiration was incorrectly calculated. Rather than setting the expiration to 1 month in the future, it set it 8 months from their previous expiration. 

The user resubscribed, and submitted payment on `2015/10/31 19:49:04`, yes expiration was changed like this:
`October 31, 2015 19:49:09 - Member's expiration changed from 2015-04-24 23:59:59 to 2015-12-03 23:59:59`

It set the month of expiration correctly, but based the calendar day and time on the old expiration.

**Comment by pippinsplugins** [comment]
@carter2422 So their account expired on October 31st and they renewed on October 31st?

**Comment by carter2422** [comment]
No, their account expired back in April, then they renewed on October 31st.

**Comment by pippinsplugins** [comment]
Are you sure their account shows as Expired and not Active or Cancelled?

**Comment by carter2422** [comment]
I'm certain. The expiration was recalculated when the resubscribed based on their old expiration date, rather than the new signup date.

**Comment by pippinsplugins** [comment]
If we look at the `calculate_expiration()` method, we can see there are two "base" points used for the expiration date calculation:

1. The member's current expiration date if they have an active subscription or their expiration date is greater than the current time, and they are not expired.

2. The current time in all other situations.

```php
/**
 * Calculates the new expiration date for a member
 *
 * @access  public
 * @since   2.4
 * @return  String Date in Y-m-d H:i:s format or "none" if is a lifetime member
*/
public function calculate_expiration() {

	// Get the member's current expiration date
	$expiration = $this->get_expiration_time();

	// Determine what date to use as the start for the new expiration calculation
	if( $expiration > current_time( 'timestamp' ) && ! $this->is_expired() && $this->get_status() == 'active' ) {

		$base_timestamp = $expiration;

	} else {

		$base_timestamp = current_time( 'timestamp' );

	}

	$subscription = rcp_get_subscription_details( $this->get_subscription_id() );

	if( $subscription->duration > 0 ) {

		$expire_timestamp  = strtotime( '+' . $subscription->duration . ' ' . $subscription->duration_unit . ' 23:59:59', $base_timestamp );
		$extension_days    = array( '29', '30', '31' );

		if( in_array( date( 'j', $expire_timestamp ), $extension_days ) && 'day' !== $subscription->duration_unit ) {

			/*
			 * Here we extend the expiration date by 1-3 days in order to account for "walking" payment dates in PayPal.
			 *
			 * See https://github.com/pippinsplugins/restrict-content-pro/issues/239
			 */

			$month = date( 'n', $expire_timestamp );

			if( $month < 12 ) {
				$month += 1;
				$year   = date( 'Y' );
			} else {
				$month  = 1;
				$year   = date( 'Y' ) + 1;
			}

			$timestamp  = mktime( 0, 0, 0, $month, 1, $year );

			$expiration = date( 'Y-m-d 23:59:59', $timestamp );
		}

		$expiration = date( 'Y-m-d 23:59:59', $expire_timestamp );

	} else {

		$expiration = 'none';

	}

	return apply_filters( 'rcp_member_calculated_expiration', $expiration, $this->ID, $this );

}
```

I'm not disputing that it happened (notes on member say otherwise), but I don't yet see where it's failing.

**Comment by pippinsplugins** [comment]
This should now be resolved with changes done for #312.
